### PR TITLE
intersection between two line segments

### LINKED ISSUE #2125: Improve intersection for line segments
Currently, `intersection` uses the default for polytopes. A new algorithm that dispatches on a pair of line segments should be implemented.

May be useful:

- http://geomalgorithms.com/a05-_intersect-1.html#Line-Intersections

### PR DESCRIPTION
**Pull request by birm**
closes #2125

Tested with

```julia
using LazySets
function i1(a::LineSegment{N}, b::LineSegment{N}) where {N<:Real}
    # cast a as line
    ap = Line(a.p, a.q)
    bp = Line(b.p, b.q)
    # find intersection between a' and b
    m = intersection(ap, bp)
    if m == ap
        # determine which segment is in both
        p1 = max(min(a.p[1], a.q[1]), min(b.p[1], b.q[1]))
        p2 = max(min(a.p[2], a.q[2]), min(b.p[2], b.q[2]))
        q1 = min(max(a.p[1], a.q[1]), max(b.p[1], b.q[1]))
        q2 = min(max(a.p[2], a.q[2]), max(b.p[2], b.q[2]))
        if p1 < q1 && p2 < q2
          return LineSegment([p1, p2], [q1, q2])
        elseif p1 == q1 && p2 == q2
          # edges have a point in common
          return Singleton([p1, p2])
        else
          # no intersection
          return return EmptySet{N}(2)
        end
    elseif m isa Singleton && m.element ∈ a && m.element ∈ b
        # if the intersection between lines is in the segments
        return m
    else
        # no intersection
        return return EmptySet{N}(2)
    end
end


## tests
s1 = LineSegment([-5.0,-5.0], [5.0,5.0])
# colinear shifted down
s2 = LineSegment([-6.0,-6.0], [4.0,4.0])
i1(s1, s2)
# parallel, not intersect
s3 = LineSegment([-5.0,-4.0], [4.0,5.0])
i1(s1, s3)
# intersect outside of segment
s4 = LineSegment([0.0,10.0], [6.0,5.0])
i1(s1, s4)
# intersect in segment
s5 = LineSegment([5.0,-5.0], [-5.0,5.0])
i1(s1, s5)
# parallel, no points in common
s6 = LineSegment([10.0, 10.0], [11.0,11.0])
i1(s1, s6)
# parallel one point in common
s7 = LineSegment([5.0,5.0], [6.0,6.0])
i1(s1, s7)


```

With output
```julia
julia> ## tests
       s1 = LineSegment([-5.0,-5.0], [5.0,5.0])
LineSegment{Float64,Array{Float64,1}}([-5.0, -5.0], [5.0, 5.0])

julia> # colinear shifted down
       s2 = LineSegment([-6.0,-6.0], [4.0,4.0])
LineSegment{Float64,Array{Float64,1}}([-6.0, -6.0], [4.0, 4.0])

julia> i1(s1, s2)
LineSegment{Float64,Array{Float64,1}}([-5.0, -5.0], [4.0, 4.0])

julia> # parallel, not intersect
       s3 = LineSegment([-5.0,-4.0], [4.0,5.0])
LineSegment{Float64,Array{Float64,1}}([-5.0, -4.0], [4.0, 5.0])

julia> i1(s1, s3)
EmptySet{Float64}(2)

julia> # intersect outside of segment
       s4 = LineSegment([0.0,10.0], [6.0,5.0])
LineSegment{Float64,Array{Float64,1}}([0.0, 10.0], [6.0, 5.0])

julia> i1(s1, s4)
EmptySet{Float64}(2)

julia> # intersect in segment
       s5 = LineSegment([5.0,-5.0], [-5.0,5.0])
LineSegment{Float64,Array{Float64,1}}([5.0, -5.0], [-5.0, 5.0])

julia> i1(s1, s5)
Singleton{Float64,Array{Float64,1}}([-0.0, 0.0])

julia> # parallel, no points in common
       s6 = LineSegment([10.0, 10.0], [11.0,11.0])
LineSegment{Float64,Array{Float64,1}}([10.0, 10.0], [11.0, 11.0])

julia> i1(s1, s6)
EmptySet{Float64}(2)

julia> # parallel one point in common
       s7 = LineSegment([5.0,5.0], [6.0,6.0])
LineSegment{Float64,Array{Float64,1}}([5.0, 5.0], [6.0, 6.0])

julia> i1(s1, s7)
Singleton{Float64,Array{Float64,1}}([5.0, 5.0])

```
I've understood the line segment to be a closed set; please let me know if you'd prefer that the s7 case isn't desired that way.

### CODE REVIEW

**Comment by schillic** [comment]
> I've understood the line segment to be a closed set; please let me know if you'd prefer that the s7 case isn't desired that way.

I agree that in this special case a `Singleton` can be returned.

The code looks good. Can you add your test cases to [`test/unit_LineSegment.jl`](https://github.com/JuliaReach/LazySets.jl/blob/cd6856388521869b89e2700393502cf326527eb3/test/unit_LineSegment.jl), please?

File: test/unit_LineSegment.jl [hunk 1]
```diff
@@ -103,6 +103,31 @@ for N in [Float64, Rational{Int}, Float32]
     intersection_empty, point = is_intersection_empty(l1, l1_copy, true)
     @test !is_intersection_empty(l1, l1_copy) && !intersection_empty && point ∈ l1
 
+    # intersection
+    s1 = LineSegment(N[-5.0,-5.0], N[5.0,5.0])
+    # colinear shifted down
+    s2 = LineSegment(N[-6.0,-6.0], N[4.0,4.0])
+    @test intersection(s1, s2) isa LineSegment
+    @test isapprox(intersection(s1, s2).p, N[-5,-5])
+    @test isapprox(intersection(s1, s2).p, N[4,4])
```
Comment (schillic) [inline test/unit_LineSegment.jl:112]: There is something wrong here. The second test fails.

Comment (birm) [reply test/unit_LineSegment.jl:112]:
```suggestion
    @test isapprox(intersection(s1, s2).p, N[-5,-5])
    @test isapprox(intersection(s1, s2).q, N[4,4])
```
little typo :sweat_smile:

File: src/ConcreteOperations/intersection.jl [hunk 2]
```diff
@@ -89,6 +89,56 @@ function intersection(L1::Line{N}, L2::Line{N}
...
+        else
+          # no intersection
+          return return EmptySet{N}(2)
+        end
+    elseif m isa Singleton && m.element ∈ a && m.element ∈ b
+        # if the intersection between lines is in the segments
+        return m
+    else
+        # no intersection
+        return return EmptySet{N}(2)
```
Comment (mforets) [inline src/ConcreteOperations/intersection.jl:138]:
```suggestion
        return EmptySet{N}(2)
```

**Comment by mforets** [comment]
Consider this case:

```julia
julia> l1 = LineSegment([0.0, 0.0], [1.0, 0.0]);
julia> l2 = LineSegment([0.0, 0.0], [2.0, 0.0]);

julia> intersection(l1, l2)
EmptySet{Float64}(2)
```
The answer should be `l1`. We can fix this behavior by using robust comparisons, perhaps swapping the order of the first two clauses. What do you think about:

```julia
...
        if LazySets._isapprox(p1, q1) && LazySets._isapprox(p2, q2)
             # edges have a point in common
             return Singleton([p1, p2])
        elseif LazySets._leq(p1, q1) && LazySets._leq(p2, q2)
             return LineSegment([p1, p2], [q1, q2])
        else
            # no intersection
            return EmptySet{N}(2)
       end
```

**Comment by birm** [comment]
Seems reasonable so done.

**Comment by birm** [comment]
@schillic if this changes your review, or if you have objections to this change, please let us know!

File: src/ConcreteOperations/intersection.jl [hunk 3]
```diff
@@ -89,6 +89,56 @@ function intersection(L1::Line{N}, L2::Line{N}
...
+        q2 = min(max(a.p[2], a.q[2]), max(b.p[2], b.q[2]))
+        if LazySets._isapprox(p1, q1) && LazySets._isapprox(p2, q2)
+             # edges have a point in common
+             return Singleton([p1, p2])
+        elseif LazySets._leq(p1, q1) && LazySets._leq(p2, q2)
+             return LineSegment([p1, p2], [q1, q2])
+        else
+            # no intersection
+            return EmptySet{N}(2)
+       end
```
Comment (schillic) [inline src/ConcreteOperations/intersection.jl:132]:
```suggestion
        end
```

**Comment by mforets** [comment]
Thanks for your contribution!

**Pull request merged**